src/votekit/ballot_generator.py

```diff
@@ -1932,53 +1932,102 @@ def generate_profile(
+    def __init__(
         self,
-        number_of_ballots: int,
-        by_bloc: bool = False,
-        dim: int = 2,
-        lower: float = 0,
-        upper: float = 1,
-        seed: Optional[int] = None,
+        candidates: list[str],
+        voter_dist: Callable[..., np.ndarray] = np.random.uniform,
+        voter_params: Optional[Dict[str, float]] = None,
+        candidate_dist: Callable[..., np.ndarray] = np.random.uniform,
+        candidate_params: Optional[Dict[str, float]] = None,
+        distance: Callable[[np.ndarray, np.ndarray], float] = euclidean_distance,
+    ):
+        super().__init__(candidates=candidates)
+        self.voter_dist = voter_dist
+        self.candidate_dist = candidate_dist
+
+        if voter_params is None:
+            # default params used for np.random.uniform
+            self.voter_params = {"low": 0.0, "high": 1.0, "size": 2.0}
+        else:
+            self.voter_params = voter_params
+
+        if candidate_params is None:
+            # default params used for np.random.uniform
+            self.candidate_params = {"low": 0.0, "high": 1.0, "size": 2.0}
+        else:
+            self.candidate_params = candidate_params
+
+        self.distance = distance
+
+    def generate_profile(
+        self, number_of_ballots: int, by_bloc: bool = False, seed: Optional[int] = None
     ) -> Union[PreferenceProfile, Tuple]:
         """
         Args:
             number_of_ballots (int): The number of ballots to generate.
             by_bloc (bool): Dummy variable from parent class.
-            dim (int, optional): number of dimensions to use, defaults to 2d
-            lower (float, optional): lower bound for uniform distribution, defaults to 0
-            upper (float, optional): upper bound for uniform distribution, defaults to 1
-            seed (int, optional): seed for random generation
+            seed (int, optional): Seed for random generation, defaults to None
+            which resets the random seed.
 
         Returns:
-            Union[PreferenceProfile, Tuple]
+            preference profile, candidate positions, voter positions:
+            (Union[PreferenceProfile, Tuple])
         """
 
         np.random.seed(seed)
         candidate_position_dict = {
-            c: np.random.uniform(lower, upper, size=dim) for c in self.candidates
+            c: self.candidate_dist(**self.candidate_params) for c in self.candidates
         }
-        voter_positions = np.random.uniform(lower, upper, size=(number_of_ballots, dim))
+        voter_positions = np.array(
+            [self.voter_dist(**self.voter_params) for v in range(number_of_ballots)]
+        )
 
         ballot_pool = []
-
-        for vp in voter_positions:
+        for v in voter_positions:
             distance_dict = {
-                c: np.linalg.norm(v - vp) for c, v, in candidate_position_dict.items()
+                i: self.distance(v, c) for i, c, in candidate_position_dict.items()
```

Comment:
This syntax is a bit confusing; save i for indices. I would do something like `cand_name, cand_coord`